### ISSUE TITLE
Allow to ignore/skip items in a playlist

### ISSUE THREAD

**Issue by tomislavturcic** [google/ExoPlayer#6170]
I use ConcatenatingMediaSource to play a track list in ExoPlayer. Some of the media sources are downloaded files. At any point during playback, user can go into Offline mode where only downloaded tracks should be played. 

Is it possible to somehow flag these downloaded media sources and make Exo skip non-flagged tracks? Or is the only solution to use concatenatingMediaSource.removeSource() and manually remove all non-downloaded sources?

I've been looking into TrackSelector class, but couldn't find anything helpful there.

**Comment by ojw28** [comment]
We don't provide this functionality currently. It's an interesting idea though, because:

* Some apps like to use ExoPlayer's playlist directly to underpin their user interface
* When going into offline mode, a developer may well just want to grey out non-playable items, rather than remove them competely

@tonihei - This seems like something we could look to provide support for at some point.

**Comment by tonihei** [comment]
There is also the enhancement tracked by #4343 which would allow to support this use case implicitly. The online sources would fail to load and would thus be skipped automatically.

More generally, we already support marking ad periods as "skipped" and the player just ignores them. A recent internal discussion about media structure shortcomings concluded that such a feature would be helpful for generic playlist items, not just ads. Let's use this to track that.

**Comment by ojw28** [comment]
> There is also the enhancement tracked by #4343 which would allow to support this use case implicitly. The online sources would fail to load and would thus be skipped automatically.

I think this is slightly different. In #4343 you'll only discover that a given item is unplayable when you try and play it (or shortly before when you start trying to buffer it). In the case described here the app might want to explicitly mark sources to be skipped, and have their UI be able to query ExoPlayer's playlist to know which items should be greyed out or displayed differently. A similar feature might be useful for marking items in a playlist that are geo-restricted or unavailable for some other reason as well. Plus having an explicit signal means an item can be skipped instantly rather than having the player attempt to play it and then fail.

**Comment by tonihei** [comment]
Yes, that's why it's not a duplicate, but a separate enhancement :)  I just wanted to point it out because both are slightly related.

**Comment by tomislavturcic** [comment]
Sorry to interject. But I have a follow-up question. 

Is there any other way to handle Online / Offline changes? Or is the only way to remove MediaSources when going to Offline mode and then add them back when going to Online mode?

And secondly, in regards to adding these new features... How long does it take to add such enhancement to releases?

**Comment by tonihei** [comment]
I'm afraid we don't provide a good way to do that currently. So yes, removing and adding them back is the only workaround I can think of at the moment.

We don't usually give time estimates for our enhancements because it's sometime unclear how complicated they are, and what things we are working on. We'll post this issue with commit or other updates once we have them.
